Lib/test/test_capi/test_tuple.py

```diff
@@ -0,0 +1,185 @@
+        self.assertRaises(SystemError, size, object())
+
+        # CRASHES size(NULL)
+
+    def test_tuple_get_size(self):
+        # Test PyTuple_GET_SIZE()
+        size = _testcapi.tuple_get_size
+
+        self.assertEqual(size(()), 0)
+        self.assertEqual(size((1, 2)), 2)
+        self.assertEqual(size(TupleSubclass((1, 2))), 2)
+
+    def test_tuple_getitem(self):
+        # Test PyTuple_GetItem()
+        getitem = _testlimitedcapi.tuple_getitem
+
+        tup = TupleSubclass((1, 2, 3))
+        self.assertEqual(getitem(tup, 0), 1)
+        self.assertEqual(getitem(tup, 2), 3)
+
+        tup = (1, 2, 3)
+        self.assertEqual(getitem(tup, 0), 1)
+        self.assertEqual(getitem(tup, 2), 3)
+
+        self.assertRaises(IndexError, getitem, tup, PY_SSIZE_T_MIN)
+        self.assertRaises(IndexError, getitem, tup, -1)
+        self.assertRaises(IndexError, getitem, tup, len(tup))
+        self.assertRaises(IndexError, getitem, tup, PY_SSIZE_T_MAX)
+        self.assertRaises(SystemError, getitem, 42, 1)
+
+        # CRASHES getitem(NULL, 1)
+
+    def test_tuple_get_item(self):
+        # Test PyTuple_GET_ITEM()
+        get_item = _testcapi.tuple_get_item
+
+        tup = TupleSubclass((1, 2, 3))
+        self.assertEqual(get_item(tup, 0), 1)
+        self.assertEqual(get_item(tup, 2), 3)
+
+        tup = (1, 2, 3)
+        self.assertEqual(get_item(tup, 0), 1)
+        self.assertEqual(get_item(tup, 2), 3)
+
+    def test_tuple_getslice(self):
+        # Test PyTuple_GetSlice()
+        getslice = _testlimitedcapi.tuple_getslice
+
+        tup = (1, 2, 3)
+
+        # empty
+        self.assertEqual(getslice(tup, PY_SSIZE_T_MIN, 0), ())
+        self.assertEqual(getslice(tup, -1, 0), ())
+        self.assertEqual(getslice(tup, 3, PY_SSIZE_T_MAX), ())
+
+        # slice
+        self.assertEqual(getslice(tup, 1, 3), (2, 3))
+
+        # whole
+        self.assertEqual(getslice(tup, 0, 3), tup)
+        self.assertEqual(getslice(tup, 0, 100), tup)
+        self.assertEqual(getslice(tup, -100, 100), tup)
+
+        # CRASHES getslice(NULL, 0, 0)
+
+    def test_tuple_setitem(self):
+        # Test PyTuple_SetItem()
+        setitem = _testlimitedcapi.tuple_setitem
+
+        tup = (0, 0)
+        self.assertEqual(setitem(tup, 0, 1), (1, 0))
+        self.assertEqual(setitem(tup, 1, 1), (0, 1))
+
+        self.assertRaises(IndexError, setitem, tup, PY_SSIZE_T_MIN, 1)
+        self.assertRaises(IndexError, setitem, tup, -1, 1)
+        self.assertRaises(IndexError, setitem, tup, len(tup), 1)
+        self.assertRaises(IndexError, setitem, tup, PY_SSIZE_T_MAX, 1)
+
+        # CRASHES setitem(NULL, 1, 5)
+
+    def test_tuple_set_item(self):
+        # Test PyTuple_SET_ITEM()
+        set_item = _testcapi.tuple_set_item
+
+        tup = (0, 0)
+        self.assertEqual(set_item(tup, 0, 1), (1, 0))
+        self.assertEqual(set_item(tup, 1, 1), (0, 1))
+
+    def test_tuple_resize(self):
+        # Test PyTuple_Resize()
```

Comment:
A more complex test is needed here. `PyTuple_Resize()` should be tested with a non-empty tuple (and also with non-tuples). It can also be used to shrink the tuple. There are also two different cases -- refcount == 1, and refcount > 1, so you need either two wrappers, or a boolean parameter which specify whether to make a fresh copy of the input tuple or use the input argument itself. See how it was done in other tests.

Reply:
I hope, I did that.  Unfortunately, it seems not too much could be done for new=False scenario.  Here we could test refcount > 1 condition.